File: backend/app/schemas/strategy.py
from datetime import datetime

from pydantic import BaseModel


class StrategyBase(BaseModel):
    type: str
    drop_percent: float
    amount_sol: float
    destination_address: str
    reference_price: float

class StrategyCreate(StrategyBase):
    cooldown_seconds: int = 60
    execution_mode: str = "recurring"
    token: str = "SOL"
    amount_usdc: float | None = None
    # swap via Jupiter
    token_in: str | None = None
    token_out: str | None = None
    slippage_bps: int | None = 50

class StrategyUpdate(BaseModel):
    type: str | None = None
    drop_percent: float | None = None
    amount_sol: float | None = None
    destination_address: str | None = None
    reference_price: float | None = None
    active: bool | None = None


class StrategyResponse(StrategyBase):
    id: int
    wallet_id: int
    active: bool
    cooldown_seconds: int
    execution_mode: str
    token: str = "SOL"
    amount_usdc: float | None = None
    token_in: str | None = None
    token_out: str | None = None
    slippage_bps: int | None = None
    last_executed_at: datetime | None
    class Config:
        from_attributes = True


class StrategyListItemResponse(StrategyResponse):
    pass

File: frontend/src/services/strategy.ts
// services/strategy.ts — versão atualizada
import { api } from "./api";

type ApiResponse<T> = { data: T; meta: any };

export const getStrategies = async () => {
  const response = await api.get<ApiResponse<Strategy[]>>("/strategies");
  return response.data.data;
};

export const createStrategy = async (payload: {
  type: string;
  drop_percent: number;
  amount_sol: number;
  destination_address: string;
  reference_price: number;
  cooldown_seconds?: number;
  execution_mode?: "once" | "recurring";
  token?: string;
  amount_usdc?: number;
  token_in?: string;
  token_out?: string;
  slippage_bps?: number;
}) => {
  const { data } = await api.post("/strategies", payload);
  return data.data;
};

// toggle ativo/inativo — chama PATCH /strategies/:id
export const toggleStrategy = async (id: number, active: boolean) => {
  const { data } = await api.patch(`/strategies/${id}`, { active });
  return data.data;
};
